### PR TITLE
Add support for singleton values

### PR DESCRIPTION
**Pull request by Kyle-Verhoog**
This allows pkgs, env and pys to have singleton values which makes the
configuration more ergonomic.

### CODE REVIEW

File: riot/riot.py [hunk 1]
```diff
@@ -24,14 +23,24 @@ def __init__(self, *args, **kwargs):
         self.__dict__ = self
 
 
+def rm_singletons(
+    d: t.Dict[str, t.Union[str, t.List[str]]]
+) -> t.Dict[str, t.List[str]]:
+    return {k: [v] if isinstance(v, str) else v for k, v in d.items()}
```
Comment (jd) [inline riot/riot.py:29]:
```suggestion
_K = t.TypeVar("_K")

def rm_singletons(
    d: t.Dict[_K, t.Union[str, t.List[str]]]
) -> t.Dict[_K, t.List[str]]:
    return {k: [v] if isinstance(v, str) else v for k, v in d.items()}
```

File: riot/riot.py [hunk 2]
```diff
@@ -24,14 +23,24 @@ def __init__(self, *args, **kwargs):
...
+
+
+def float_to_list(x: t.Union[float, t.List[float]]) -> t.List[float]:
+    return [x] if isinstance(x, float) else x
```
Comment (jd) [inline riot/riot.py:33]:
```suggestion
_K = t.TypeVar("_K")
def to_list(x: t.Union[_K, t.List[_K]]) -> t.List[_K]:
    return [x] if not isinstance(x, collections.abc.Iterable) else x
```

and use it in `rm_singletons`

File: tests/test_cli.py [hunk 3]
```diff
@@ -385,3 +385,54 @@ def test_failure():
         result = cli.invoke(riot.cli.main, ["run", "-s", "failure"])
         assert result.exit_code == 1
         assert result.stdout == ""
+
+
+def test_types(cli: click.testing.CliRunner):
```
Comment (jd) [inline tests/test_cli.py:390]:
```suggestion
def test_types(cli: click.testing.CliRunner) -> None:
```

**Review by jd — CHANGES_REQUESTED** [review]

File: riot/riot.py (hunk 2 above)
Comment (Kyle-Verhoog) [reply riot/riot.py:33]: Hmm I had tried something like this initially and had gotten:

image

I wasn't using `collections.abc.Iterable` though, so maybe that was it!

**Comment by Kyle-Verhoog** [comment]
Yeah @jd getting the same type error as I got before:


image


Not quite sure how to resolve 🤔

File: riot/riot.py [hunk 4]
```diff
@@ -24,14 +23,27 @@ def __init__(self, *args, **kwargs):
...
 
+_K = t.TypeVar("_K")
+
+def rm_singletons(
+    d: t.Dict[_K, t.Union[str, t.List[str]]]
+) -> t.Dict[_K, t.List[str]]:
+    return {k: [v] if isinstance(v, str) else v for k, v in d.items()}
+
+
+_K = t.TypeVar("_K")
```
Comment (jd) [inline riot/riot.py:34]: You don't need to redefine it here, you can use the same one.

File: riot/riot.py [hunk 5]
```diff
@@ -32,15 +31,29 @@ def __init__(self, *args, **kwargs):
         self.__dict__ = self
 
 
+def rm_singletons(d: t.Dict[_K, t.Union[str, t.List[str]]]) -> t.Dict[_K, t.List[str]]:
+    return {k: [v] if isinstance(v, str) else v for k, v in d.items()}
```
Comment (jd) [inline riot/riot.py:35]:
```suggestion
def rm_singletons(d: t.Dict[_K, t.Union[_V, t.List[_V]]]) -> t.Dict[_K, t.List[_V]]:
    return {k: to_list(v) for k, v in d.items()}
```

**Review by jd — CHANGES_REQUESTED** [review]

File: riot/riot.py [hunk 6]
```diff
@@ -32,15 +31,29 @@ def __init__(self, *args, **kwargs):
         self.__dict__ = self
 
 
+def rm_singletons(d: t.Dict[_K, t.Union[_V, t.List[_V]]]) -> t.Dict[_K, t.List[_V]]:
+    return {k: to_list(v) for k, v in d.items()}
+
+
+def to_list(x: t.Union[_K, t.List[_K]]) -> t.List[_K]:
+    return [x] if not isinstance(x, list) else x
```
Comment (jd) [inline riot/riot.py:39]: I'm approving this but having 2 simple unit tests for those functions could be worth the 5 minutes.
Feel free to add them later.

**Review by jd — APPROVED** [review]

File: riot/riot.py (hunk 6 above)
Comment (Kyle-Verhoog) [reply riot/riot.py:39]: seems like a good place for some doctests!

Comment (Kyle-Verhoog) [reply riot/riot.py:39]: Done in https://github.com/DataDog/riot/pull/77

**Pull request merged**